File: framework/languages/java/samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/adapters/zlink/actors/SupportActorDirectory.java
package systems.zlink.samples.supportchat.server.support.adapters.zlink.actors;

import java.util.HashMap;
import java.util.Map;

public final class SupportActorDirectory {
    private final Object gate = new Object();
    private final Map<String, SupportUserActor> actors = new HashMap<>();

    public void addOrUpdate(SupportUserActor actor) {
        synchronized (gate) {
            actors.put(actor.actorId(), actor);
        }
    }

    public SupportUserActor get(String actorId) {
        synchronized (gate) {
            SupportUserActor actor = actors.get(actorId);
            if (actor == null) {
                throw new IllegalStateException("Support actor is not available. actor=" + actorId);
            }
            return actor;
        }
    }
}

File: framework/languages/java/samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/adapters/zlink/actors/SupportActorDirectory.kt
package systems.zlink.samples.kotlin.supportchat.server.support.adapters.zlink.actors

class SupportActorDirectory {
    private val gate = Any()
    private val actors = mutableMapOf<String, SupportUserActor>()

    fun addOrUpdate(actor: SupportUserActor) {
        synchronized(gate) {
            actors[actor.actorId()] = actor
        }
    }

    fun get(actorId: String): SupportUserActor =
        synchronized(gate) {
            actors[actorId]
                ?: throw IllegalStateException("Support actor is not available. actor=$actorId")
        }
}
